# Sample Data

Sample input data for scripts, Used for testing.

- `sample_exogeneous_sequence/`: sample files representing 
  signal tracks for exogeneous sequences, including a fasta 
  file for sequences and 2 npy files for fwd and reverse 
  strand signals. The data is generated by overlapping 
  WTC11-derived Microglias' distal PROcap elements with 
  (Bellenguez, 2022) and (Wightman, 2021) GWAS hits 
  (cutoff: 5e-8). Signal tracks are predicted by a 
  MG-trained PROcapNet model.

- `sample_rsem_results`: rsem result file from 2 random 
  samples.

- `ENCFF182TPF.mn.bigWig`: bigwig signal track for K562 
  PROcap from ENCODE accession ENCFF182TPF.

- `ENCFF993VCR.pl.bigWig`: bigwig signal track for K562 
  PROcap from ENCODE accession ENCFF993VCR.

- `ENCFF819CPA.K562_PROcap.bidirectionalPeak.bedtre`: 
  PINTS call bed TRE file for K562 from ENCODE 
  accession ENCFF819CPA.

- `fastp.json`: fastp json file for a random 
  sequencing library.

- `gencode.v37.gene_id.list`: List file for gene ids 
  from gencode v37.

- `gencode.v37.protein_coding_gene_body.tss.chr2.bed6`: 
  All the protein coding gene tss in chr2 from GENCODE 
  v37 in bed6 format.

- `sample.bam_flagstat.txt`: bam_flagstat file for a random 
  sequencing library.

- `sample.bedpe`: An example bedpe file

- `sample.chr2.snp.bed`: An example snp bed file. Generated from 
  all the chr2 significant hits from (Bellenguez, 2022).

- `sample.GENCODE.gtf`: sample GENCODE format gtf file. generated 
  from all the gtf entries related to BIN1 from GENCODE v37.

- `sample.gwas_summary.bed6`: sample GWAS hits bed6 file. Generated 
  from all the BIN1 proximal SNPs from (Bellenguez, 2022).

- `sample.snp_overlapping_tre.chr2.bed`: Example tre bed3 file. 
  Generated with all the (Bellenguez, 2022) SNP-overlapping TREs in MG.

- `sample.STAR_log`: STARR aligner log example.

- `sample1.annot.bed3`: example annotation bed3 file.
